### PR TITLE
[SYCL-MLIR] Allow result type of sycl method to be pointer of struct type

### PR DESCRIPTION
**Pull request by whitneywhtsang**
Pointer to struct type is represented as `!llvm.ptr<struct<...>>`. 
The result type of e.g., `sycl.accessor.subscript` can be of `!llvm.ptr` type, if the accessor is to a struct type.

Signed-off-by: Tsang, Whitney <whitney.tsang@intel.com>

### CODE REVIEW

File: mlir-sycl/lib/Dialect/IR/SYCLOps.cpp [hunk 1]
```diff
@@ -98,18 +98,25 @@ mlir::LogicalResult mlir::sycl::SYCLAccessorSubscriptOp::verify() {
...
       return emitOpError(
                  "Expecting a reference to this accessor's value type (")
-             << AccessorTy.getType() << "). Got " << resultType;
+             << AccessorTy.getType() << "). Got " << ResultType;
     }
 
+    if (PT && !ElemType.isa<LLVM::LLVMStructType>())
+      return emitOpError("Expecting pointer to struct return type. Got ")
+             << ResultType;
+
```
Comment (victor-eds) [inline mlir-sycl/lib/Dialect/IR/SYCLOps.cpp:119]: NIT: This LGTM, but maybe it'd be cleaner with a TypeSwitch with a single case handling both MemRef and pointer (last if would be `if constexpr (std::is_same<decltype(PT), LLVM::LLVMStructType>)` and we would not need to evaluate at runtime.

**Review by victor-eds — APPROVED** [review]

File: mlir-sycl/lib/Dialect/IR/SYCLOps.cpp (hunk 1 above)
Comment (whitneywhtsang) [reply mlir-sycl/lib/Dialect/IR/SYCLOps.cpp:119]: The last if check if the element type of the llvm pointer is a struct type, we don't know the element type at compile time.

Comment (victor-eds) [reply mlir-sycl/lib/Dialect/IR/SYCLOps.cpp:119]: True, we'd need to check that the input type is a pointer (at compile time) and the element type is a struct type in that case (at runtime)

Comment (whitneywhtsang) [reply mlir-sycl/lib/Dialect/IR/SYCLOps.cpp]: Done.

File: mlir-sycl/lib/Dialect/IR/SYCLOps.cpp [hunk 2]
```diff
@@ -98,19 +98,31 @@ mlir::LogicalResult mlir::sycl::SYCLAccessorSubscriptOp::verify() {
...
+            return emitOpError(
+                       "Expecting a reference to this accessor's value type (")
+                   << AccessorTy.getType() << "). Got " << Ty;
+
+          if constexpr (std::is_same<decltype(Ty),
+                                     LLVM::LLVMPointerType>::value) {
+            if (!ElemType.isa<LLVM::LLVMStructType>())
+              return emitOpError(
+                         "Expecting pointer to struct return type. Got ")
+                     << ResultType;
```
Comment (etiotto) [inline mlir-sycl/lib/Dialect/IR/SYCLOps.cpp:117]: I would separate the Case into one for memref and one for LLVM pointer (given this code is only for LLVM ptr). 
Code at lines 106-110 would be duplicated, but you can make that into a lambda and call the lambda function in 2 places. Suggesting `CheckElemType` for the lambda name.

File: mlir-sycl/lib/Dialect/IR/SYCLOps.cpp [hunk 3]
```diff
@@ -98,19 +98,31 @@ mlir::LogicalResult mlir::sycl::SYCLAccessorSubscriptOp::verify() {
...
-    if (resultType.getElementType() != AccessorTy.getType()) {
-      return emitOpError(
-                 "Expecting a reference to this accessor's value type (")
-             << AccessorTy.getType() << "). Got " << resultType;
-    }
-
-    return success();
+          return success();
+        })
+        .Default([&](auto Ty) {
```
Comment (etiotto) [inline mlir-sycl/lib/Dialect/IR/SYCLOps.cpp:122]: [&] -> [ ]  (or [this] if emitOpError is a member function

File: polygeist/tools/cgeist/Test/Verification/sycl/functions.cpp [hunk 4]
```diff
@@ -76,6 +76,21 @@ SYCL_EXTERNAL void accessor_subscript_operator_2(sycl::accessor<sycl::cl_int, 1>
   keep(acc[index]);
 }
 
+// CHECK-MLIR-LABEL: func.func @_Z29accessor_subscript_operator_3N4sycl3_V18accessorI6StructLi1ELNS0_6access4modeE1026ELNS3_6targetE2014ELNS3_11placeholderE0ENS0_3ext6oneapi22accessor_property_listIJEEEEEm(
+// CHECK-MLIR:          %{{.*}}: memref<?x!sycl_accessor_1_21llvm2Estruct3C28i32293E_rw_gb> {llvm.align = 8 : i64, llvm.byval = !sycl_accessor_1_21llvm2Estruct3C28i32293E_rw_gb, llvm.noundef}, 
+// CHECK-MLIR-SAME:     %{{.*}}: i64 {llvm.noundef})
+// CHECK-MLIR: %{{.*}} = sycl.accessor.subscript %{{.*}}[%{{.*}}] {BaseType = memref<?x!sycl_accessor_1_21llvm2Estruct3C28i32293E_rw_gb, 4>, FunctionName = @"operator[]", MangledFunctionName = @_ZNK4sycl3_V18accessorI6StructLi1ELNS0_6access4modeE1026ELNS3_6targetE2014ELNS3_11placeholderE0ENS0_3ext6oneapi22accessor_property_listIJEEEEixILi1EvEERS2_NS0_2idILi1EEE, TypeName = @accessor} : (!sycl_accessor_1_21llvm2Estruct3C28i32293E_rw_gb, memref<?x!sycl_id_1_>) -> !llvm.ptr<struct<(i32)>, 4>
```
Comment (etiotto) [inline polygeist/tools/cgeist/Test/Verification/sycl/functions.cpp:82]: Where is sycl_accessor_1_21llvm2Estruct3C28i32293E_rw_gb declared ?

**Pull request merged**